### PR TITLE
Docs: Add docs on services

### PR DESCRIPTION
**Pull request by marcusolsson**
**What this PR does / why we need it**:

This pull request adds documentation for the backend services.

### CODE REVIEW

File: contribute/architecture/services.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+Even though the services in Grafana do different things, they share a number of patterns. To better understand how a service works, let's build one from scratch!
+
+### Create a service
+
+To start building a service:
+
+- Create a new Go package `mysvc` in the [pkg/services](/pkg/services) directory.
+- Create a `service.go` file inside your new directory.
+
+All services need to implement the [Service](https://godoc.org/github.com/grafana/grafana/pkg/registry#Service) interface:
```
Comment (gotjosh) [inline contribute/architecture/services.md]: 🎉

File: contribute/architecture/services.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+```go
+type MyService struct {
+}
+
+func (s *MyService) Init() error {
+    return nil
+}
+```
+
+The `Init` method is used to initialize and configure the service to make it ready to use.
```
Comment (gotjosh) [inline contribute/architecture/services.md]:
```suggestion
The `Init` method is used to initialize and configure the service to make it ready to use. Services that return an error will halt Grafana's startup process and cause the error to be logged as it exits.
```

File: contribute/architecture/services.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    Count:   1,
+}
+if err := s.bus.Publish(event); err != nil {
+    return err
+}
+```
+
+### Commands
+
+A command is a request for an change to be made. Unlike an event, a command can be declined by the handler for that command. The handler will then return an error.
```
Comment (gotjosh) [inline contribute/architecture/services.md]: I love the fact that we're properly defining what a command is - however, not sure this is what the definition should look like. Commands are not always _changes_ in the system. The way I see it they are just an action to be handled.

On another note (potential dumb question),
>a command can be declined by the handler for that command

I'm not sure what this means in this context. How can the handler decline the execution of a command? Is the case where a handler is not present for the command?

File: contribute/architecture/services.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+> Because we request an operation to be performed, command are written in imperative mood, such as `CreateFolderCommand`, and `DeletePlaylistCommand`.
+
+#### Dispatch a command
+
+To dispatch a command to another service, pass a command object to the `Dispatch` method:
```
Comment (gotjosh) [inline contribute/architecture/services.md]:
```suggestion
To dispatch a command, pass the object to the `Dispatch` method:
```

**Review by gotjosh — COMMENTED** [review]
There's plenty of wording under the assumption that the bus's main purpose is to communicate between services - I'm not sure that's the case. I'll be happy to take a second of look once I'm clearer with that assumption.

File: contribute/architecture/services.md (hunk 2 above)
Comment (marcusolsson) [reply contribute/architecture/services.md]:
```suggestion
The `Init` method is used to initialize and configure the service to make it ready to use. Services that return an error halts Grafana's startup process and cause the error to be logged as it exits.
```

Minor change to keep consistent present tense.

File: contribute/architecture/services.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+func (s *MyService) Init() error {
+    return nil
+}
+```
+
+The `Init` method is used to initialize and configure the service to make it ready to use.
+
+### Register a service
+
+Every service needs to be registered for the application to find it.
```
[the remark this answers is not in the document]
Comment (marcusolsson) [reply contribute/architecture/services.md]: Agreed. Though I should probably explain what a service graph is, in that case :)

File: contribute/architecture/services.md (hunk 3 above)
Comment (marcusolsson) [reply contribute/architecture/services.md]: Would "request for an action to be taken" work?

I'm trying to convey that unlike events, command are not facts. It's up to the handler to decide whether the command is valid, and whether and any action is taken/change is made. I've seen implementations (outside the Grafana project) where commands are handled as events, which will result in inconsistencies if the handler wasn't able to complete the command.

By decline, I'm referring to handler logic that might decide that the system isn't able to complete the command. For example, if a `DeleteDashboardCommand` is dispatched, but no dashboard with that ID is available, the handler can't fulfill the request. 

Is there a better way to phrase this?

Comment (gotjosh) [reply contribute/architecture/services.md]: > Would "request for an action to be taken" work?

Love it, let's run with that ❤️ 

> Is there a better way to phrase this?

That makes sense! How do you feel about: 

`Unlike an event's fire and forget approach, a command can fail as by its handler.`

Declination of a handler would imply that the command was never received by the handler as opposed to a failure to complete execution.

**Review by gotjosh — APPROVED** [review]
img

Fantastic work 🥇, as always all my comments are optional. I don't need to see this again as I'm very happy with the current result - merge at will.

File: contribute/architecture/communication.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    Count:   1,
+}
+if err := s.bus.Publish(event); err != nil {
+    return err
+}
+```
+
+## Commands
+
+A command is a request for an actions to be taken. Unlike an event's fire-and-forget approach, a command can fail as it is handled. The handler will then return an error.
```
[the remark this answers is not in the document]
Comment (marcusolsson) [reply contribute/architecture/communication.md]:
```suggestion
A command is a request for an action to be taken. Unlike an event's fire-and-forget approach, a command can fail as it is handled. The handler will then return an error.
```

**Review by xlson — APPROVED** [review]
Great work on documenting this! <3

**Pull request merged**